Blog post: Powering Functions at the Edge with OpenFaaS (#275)
openfaas/openfaas.github.io: _posts/2022-05-27-pro-components-with-faasd.md

---

> title: "Exploring OpenFaaS Pro without Kubernetes"

Reviewer: What other titles could you come up with?

Reviewer: The Event-driven Edge with OpenFaaS

Reviewer 2: Handle Events at the Edge with OpenFaaS

description: "In this post Han shows how the OpenFaaS Pro components can be used with faasd."
date: 2022-05-27

> categories:

Reviewer: I make these lowercase usually

  - Faasd
  - Configuration
  - Kafka
author_staff_member: han
---

In this post Han shows how the OpenFaaS Pro components can be used with faasd.

## What is faasd?
Faasd is the lightweight option to run OpenFaas. It does not require Kubernetes but uses containerd and runc instead. Beacause of this you can run faasd almost anywhere.

Faasd comes with all the core components from the OpenFaaS stack. It has a REST management api and ui dashboard. It comes with a queue system built around NATS for asynchronous invocations and has built in Prometheus monitoring.

While this is already a strong set of features, if you are using faasd in production you might want to extend its feature set with some of the OpenFaaS Pro components. These features include:
- An extended UI dashboard
- Event connectors to trigger functions from Kafka or AWS SQS events
- Scale to zero
- Retries
- Single Sign-On

## Why faasd?

> Faasd uses containerd directly, this eliminates the need to learn and operate Kubernetes. Leaving out Kubernetes also has the advantage that faasd can be deployed on devices that have limited resources. This makes it ideal for use at the edge, on IoT devices and in industrial or airgapped environments.

Reviewer: Faasd => faasd. First one could be a link to GitHub?

Faasd works well for the same kind of use-cases as OpenFaaS on Kubernetes. It gives you a universal API to deploy microservices or functions written in any language [using our templates or by creating your own](https://docs.openfaas.com/cli/templates/#templates). Some of these use-cases may include:

- An integration for a data-feed from third-parties.

> - An API written in Pyhton, ASP.net, Node.js or any other language.

Reviewer: = > Python

> - An API written in Pyhton, ASP.net, Node.js or any other language.

Reviewer: ASP.net => "Dot net core"

Reviewer: Go?

- Running batch jobs.
- Receiving webhooks.
- 

## Getting started

> The official guide to faasd, [Serverless For Everyone Else](https://openfaas.gumroad.com/l/serverless-for-everyone-else) is the best way te learn about faasd and OpenFaaS.

Reviewer 2: ```suggestion
The official guide to faasd, [Serverless For Everyone Else](https://openfaas.gumroad.com/l/serverless-for-everyone-else) is the best way to learn about faasd and OpenFaaS.
```

There are also some tutorials that show you how to deploy faasd on different platforms:

- [Deploy faasd to DigitalOcean with Terraform](https://www.openfaas.com/blog/faasd-tls-terraform/)
- [Deploy faasd to your laptop with Multipass](https://github.com/openfaas/faasd/blob/master/docs/MULTIPASS.md)

## Deploying additional services on faasd
By default faasd runs a stack of services that make up the core of OpenFaaS. It uses a subset of directives from the [compose-spec](https://compose-spec.io/) to define these services in a docker-compose.yaml file. You can extend this file to deploy additional services.

> > The book, [Serverless For Everyone Else](https://openfaas.gumroad.com/l/serverless-for-everyone-else), has a complete chapter on how to add additional services and containers to faasd. It includes examples for running Grafana, Potgresql, Influxdb and Redis.

Reviewer 2: ```suggestion
> The book, [Serverless For Everyone Else](https://openfaas.gumroad.com/l/serverless-for-everyone-else), has a complete chapter on how to add additional services and containers to faasd. It includes examples for running Grafana, PostgreSQL, Influxdb and Redis.
```

In order to run the pro components, an [OpenFaaS Pro license](https://www.openfaas.com/support/) is required. This license file will have to be created on the faasd host at `/var/lib/faasd/secrets/openfaas-license`.

## The dashboard

> OpenFaaS recently released a new UI Dashboard. The dashboard offers you a nice visual way te discover deployed functions. It alows you to see the deployed functions per namespace and shows you the total invocation count and replica count of each function along with some other metadata.

Reviewer 2: ```suggestion
OpenFaaS recently released a new UI Dashboard. The dashboard offers you a nice visual way to discover deployed functions. It enables you to see the deployed functions per namespace and shows you the total invocation count and replica count of each function along with other rich metadata.
```

To deploy the faas-idler add the following config to your `docker-compose.yaml` file:
```yaml
faas-idler:
    image: "ghcr.io/openfaasltd/faas-idler:0.5.0"
    environment:
      - "gateway_url=http://gateway:8080"
      - "secret_mount_path=/run/secrets/"
      - "prometheus_host=prometheus"
      - "prometheus_port=9090"
      # If a function is inactive for x minutes, it may be scaled to zero
      - "inactivity_duration=3m" 
      # The interval between each attempt to scale functions to zero
      - "reconcile_interval=2m" 
      # Write additional debug information
      - "write_debug=false"
    command:
      - "/usr/bin/faas-idler"
      - "-license-file=/run/secrets/openfaas-license"
    volumes:
      - type: bind
        source: "./secrets/openfaas-license"
        target: "/run/secrets/openfaas-license"
      - type: bind
        source: ./secrets/basic-auth-password
        target: /run/secrets/basic-auth-password
      - type: bind
        source: ./secrets/basic-auth-user
        target: /run/secrets/basic-auth-user
    depends_on:
      - gateway
      - prometheus
    cap_add:
      - CAP_NET_RAW
```

> One of the strenghts of faasd compared to OpenFaaS on Kubernetes is its sub second cold-start time. On Kubernetes scaling up form zero can take a couple of seconds. Faasd directly uses containerd which significantly reduces this cold start period.

Reviewer 2: ```suggestion
One of the strengths of faasd compared to OpenFaaS on Kubernetes is its sub-second cold start time. On Kubernetes scaling up from zero can take a couple of seconds. In comparison, faasd uses containerd directly, which significantly reduces this cold start period.
```

We can deploy the figlet function from the OpenFaaS store and add the required label.  
```
faas-cli store deploy figlet \
  --label com.openfaas.scale.zero=true
```

After leaving the function idle for the configured time period it will be scaled to zero replicas.
```
Function                        Invocations     Replicas
figlet                          1               0
nodeinfo                        0               1 
```

When we call the function after it was scaled down the container will be restarted and once it is ready the function will be called. This all happens in under 90ms indluding the round trip latency.

```
time curl http://127.0.0.1:8080/function/figlet -d "Scale"
 ____            _      
/ ___|  ___ __ _| | ___ 
\___ \ / __/ _` | |/ _ \
 ___) | (_| (_| | |  __/
|____/ \___\__,_|_|\___|
                        

real    0m0.085s
```

## Retrying requests

> It is possible to limit the amount of concurrent request that your function can process. This can be usfull when you are running memory intensive tasks and want to prevent overloading the function. Limiting requests can be done by setting `max_inflight` on the function. When a function cannot accept any more connections it will return a http status code 429. You could handle these types of errors by implementing your own retry machnism which but that would probabily be a tedious process. OpenFaaS Pro comes with an upgraded queue-worker that will retry messages a number of times using an exponential back-off arlgorithm. The pro queue-worker will retry messages a set number of times before dropping them. 

Reviewer 2: ```suggestion
It is possible to limit the number of concurrent request that your function can process. This can be useful when you are running memory intensive tasks and want to prevent overloading the function. Limiting requests can be done by setting `max_inflight` on the function. When a function cannot accept any more connections it will return a http status code 429. You could handle these types of errors by implementing your own retry mechanism which would probably be a time consuming and tedious process. OpenFaaS Pro offers an upgraded queue-worker that will retry messages a number of times using an exponential back-off algorithm. The pro queue-worker will retry messages a set number of times before dropping them. 
```

```
export OPENFAAS_URL=""
export DOCKER_USER=""

faas-cli new --lang python3 user-fn
```

Our function has some third party python dependencies that we will need to add to the requirements file.

Edit `./user-fn/requirements.txt` and add this line:
```
Jinja2
```

Now we can edit the function handler `./user-fn/handler.py`:
```python
from jinja2 import Template
import json
import sys

def handle(req):
    input = json.loads(req) 

    t = Template("User: {{id}} Time: {{time}} Gender: {{gender}}")
    res = t.render(id=input["userid"], time=input["registertime"], gender=input["gender"])
    return res 
```

Before we deploy our function we will add an annotation for the `user.signup` topic so that the function is invoked for any message received.

Edit `./user-fn.yml`: 
```yaml
version: 1.0
provider:
  name: openfaas
  gateway: http://127.0.0.1:8080
functions:
  user-fn:
    annotations:
      topic: user.signup
    lang: python3
    handler: ./user-fn
    image: user-fn:latest
```

We can now build and deploy the function.
```
faas-cli up -f user-fn.yml
```

> When you look at the function detail of the user-fn it the dashboard you should see that the function is being invoked successfully. The logs of the kafka-connector can also help to see if the function is consuming messages.

Reviewer 2: `user-fn it the` => `user-fn in the`

![Logs of the kafka connector that show messages being consumed by the user-fn function](/images/2022-05-faasd-pro/kafka-connector-logs.png)
> Logs of the kafka connector that show messages being consumed by the user-fn function.

![Dashboard view of user-fn showing a rising invocation count](/images/2022-05-faasd-pro/kafka-connector-dashboard.png)
> Dashboard view of user-fn showing a rising invocation count.

# Wrapping up

> If running OpenFaaS on Kubernetes is not an option for you because your devices have limited resources or you want to avoid the complexity of running and managing Kubernetes, faasd can be the perfect alternative. It supports most of the OpenFaaS features and can be extended with OpenFaaS Pro features. 

Reviewer: "perfect" may be too strong / subjective
